File: android/app/src/main/kotlin/com/bioliminal/app/pose/PosePlugin.kt
package com.bioliminal.app.pose

import android.content.Context
import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Flutter plugin bridging Dart → native MediaPipe Tasks PoseLandmarker.
 *
 * Channel: `bioliminal.app/pose`. Inference runs on a single-threaded
 * background executor; concurrent calls drop the new frame (defense in
 * depth — Dart side also drops while busy).
 */
class PosePlugin : FlutterPlugin, MethodCallHandler {

    companion object {
        private const val CHANNEL_NAME = "bioliminal.app/pose"
    }

    private var channel: MethodChannel? = null
    private var context: Context? = null
    private var flutterAssets: FlutterPlugin.FlutterAssets? = null
    private var helper: PoseLandmarkerHelper? = null
    private val executor: ExecutorService = Executors.newSingleThreadExecutor()
    private val isProcessing = AtomicBoolean(false)
    private val mainHandler = Handler(Looper.getMainLooper())

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        flutterAssets = binding.flutterAssets
        channel = MethodChannel(binding.binaryMessenger, CHANNEL_NAME).apply {
            setMethodCallHandler(this@PosePlugin)
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        executor.shutdown()
        helper?.close()
        helper = null
        channel?.setMethodCallHandler(null)
        channel = null
        context = null
        flutterAssets = null
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "initialize" -> handleInitialize(call, result)
            "processFrame" -> handleProcessFrame(call, result)
            "dispose" -> handleDispose(result)
            else -> result.notImplemented()
        }
    }

    private fun handleInitialize(call: MethodCall, result: Result) {
        val ctx = context
        val assets = flutterAssets
        val assetPath = call.argument<String>("assetPath")
        if (ctx == null || assets == null || assetPath == null) {
            result.success(false)
            return
        }
        // Resolve Flutter asset to the path MediaPipe can read via AssetManager.
        val resolved = assets.getAssetFilePathByName(assetPath)
        executor.submit {
            try {
                val newHelper = PoseLandmarkerHelper(ctx).apply { setup(resolved) }
                helper?.close()
                helper = newHelper
                replyOnMain(result, true)
            } catch (e: Exception) {
                replyOnMain(result, false)
            }
        }
    }

    private fun handleProcessFrame(call: MethodCall, result: Result) {
        val helper = this.helper
        if (helper == null) {
            result.success(emptyList<Map<String, Double>>())
            return
        }
        if (!isProcessing.compareAndSet(false, true)) {
            result.success(emptyList<Map<String, Double>>())
            return
        }

        val bytes = call.argument<ByteArray>("bytes")
        val width = call.argument<Int>("width")
        val height = call.argument<Int>("height")
        val rotationDegrees = call.argument<Int>("rotationDegrees")
        val timestampMs = (call.argument<Number>("timestampMs"))?.toLong()

        if (bytes == null || width == null || height == null ||
            rotationDegrees == null || timestampMs == null
        ) {
            isProcessing.set(false)
            result.success(emptyList<Map<String, Double>>())
            return
        }

        executor.submit {
            try {
                val landmarks = helper.detectForVideo(
                    bytes, width, height, rotationDegrees, timestampMs,
                )
                replyOnMain(result, landmarks)
            } catch (_: Exception) {
                replyOnMain(result, emptyList<Map<String, Double>>())
            } finally {
                isProcessing.set(false)
            }
        }
    }

    private fun handleDispose(result: Result) {
        executor.submit {
            helper?.close()
            helper = null
            replyOnMain(result, null)
        }
    }

    private fun replyOnMain(result: Result, value: Any?) {
        mainHandler.post { result.success(value) }
    }
}

File: ios/Runner/Pose/PosePlugin.swift
import Flutter
import Foundation

/// Flutter plugin bridging Dart → native MediaPipe Tasks PoseLandmarker.
///
/// Channel: `bioliminal.app/pose`. Inference runs on a serial background
/// queue; concurrent frames are dropped via an atomic flag (defense in
/// depth — Dart side also drops while busy).
public class PosePlugin: NSObject, FlutterPlugin {

    private static let channelName = "bioliminal.app/pose"

    private let helper = PoseLandmarkerHelper()
    private let queue = DispatchQueue(label: "com.bioliminal.app.pose", qos: .userInitiated)
    private var isProcessing = false
    private let lock = NSLock()
    private weak var registrar: FlutterPluginRegistrar?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: channelName,
            binaryMessenger: registrar.messenger()
        )
        let instance = PosePlugin(registrar: registrar)
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    init(registrar: FlutterPluginRegistrar) {
        self.registrar = registrar
        super.init()
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize": handleInitialize(call, result: result)
        case "processFrame": handleProcessFrame(call, result: result)
        case "dispose": handleDispose(result: result)
        default: result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Methods

    private func handleInitialize(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let registrar = self.registrar,
              let args = call.arguments as? [String: Any],
              let assetPath = args["assetPath"] as? String else {
            DispatchQueue.main.async { result(false) }
            return
        }
        let delegate = (args["delegate"] as? String) ?? "cpu"
        let key = registrar.lookupKey(forAsset: assetPath)
        guard let modelPath = Bundle.main.path(forResource: key, ofType: nil) else {
            DispatchQueue.main.async { result(false) }
            return
        }
        queue.async { [weak self] in
            guard let self = self else { return }
            do {
                try self.helper.setup(modelAssetPath: modelPath, delegate: delegate)
                DispatchQueue.main.async { result(true) }
            } catch {
                DispatchQueue.main.async { result(false) }
            }
        }
    }

    private func handleProcessFrame(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let bytesData = args["bytes"] as? FlutterStandardTypedData,
              let width = args["width"] as? Int,
              let height = args["height"] as? Int,
              let bytesPerRow = args["bytesPerRow"] as? Int,
              let rotationDegrees = args["rotationDegrees"] as? Int,
              let timestampMs = args["timestampMs"] as? Int else {
            result([] as [[String: Double]])
            return
        }

        lock.lock()
        if isProcessing {
            lock.unlock()
            result([] as [[String: Double]])
            return
        }
        isProcessing = true
        lock.unlock()

        let bgra = bytesData.data
        queue.async { [weak self] in
            guard let self = self else { return }
            let landmarks = self.helper.detectForVideo(
                bgra: bgra,
                width: width,
                height: height,
                bytesPerRow: bytesPerRow,
                rotationDegrees: rotationDegrees,
                timestampMs: timestampMs
            )
            self.lock.lock()
            self.isProcessing = false
            self.lock.unlock()
            DispatchQueue.main.async { result(landmarks) }
        }
    }

    private func handleDispose(result: @escaping FlutterResult) {
        queue.async { [weak self] in
            self?.helper.close()
            DispatchQueue.main.async { result(nil) }
        }
    }
}
